### PR TITLE
test/implement end2end for m2

### PR DESCRIPTION
**Pull request by Alex720p**
This pr aims to implement an e2e test for the M2 deliverable.

- **`End2EndGeneral.kt`**
  - Implement `e2eNavigationAndDisplaysCorrectDetailsExceptForParks` e2e test that tests the progression UI and the Profile UI.

- **`MainActivity.kt`**
  - Add parameter `navTestInvokationOnEachRecompose` to `StreetWorkApp` composable, that allows for backward compability in older tests as the navTestInvokation blocked navigation in end2end testing.

- **`BottomNavigationMenu.kt`**
  - Rename testTag `bottomNavigationItem` to `bottomNavigationItem${Route}` for easier navigation flow in end2end.

- **`BottomNavavigationTest.kt` & `TopAppBarTest.tk`**
  - Add parameter `navTestInvokationOnEachRecompose` that allows for backward compability in older tests, as the navTestInvokation blocked navigation in end2end testing.

**Note:** Ignore all files that are related to Progression, I had to merge the branch from pr https://github.com/SwEnt-Group8/Street-work-app/pull/93 into this one to integrate the end2end.

### CODE REVIEW

File: app/src/androidTest/java/com/android/streetworkapp/ui/Progression/ProgressionTest.kt
Comment (Simmanz) [inline app/src/androidTest/java/com/android/streetworkapp/ui/Progression/ProgressionTest.kt:1]: Is it a good idea to add this file to your PR ? From my understanding it is about implementing the end2end test and not about the other tests ?

File: app/src/androidTest/java/com/android/streetworkapp/ui/navigation/BottomNavigationTest.kt
Comment (Simmanz) [inline app/src/androidTest/java/com/android/streetworkapp/ui/navigation/BottomNavigationTest.kt:1]: Same question as before, should we really have test changes in this PR.

File: app/src/androidTest/java/com/android/streetworkapp/ui/navigation/TopAppBarTest.kt
Comment (Simmanz) [inline app/src/androidTest/java/com/android/streetworkapp/ui/navigation/TopAppBarTest.kt:1]: Same question here, should we really have test changes in this PR

File: app/src/main/java/com/android/streetworkapp/ui/progress/Progression.kt
Comment (Simmanz) [inline app/src/main/java/com/android/streetworkapp/ui/progress/Progression.kt:1]: What is the point of modifying your UI screen this much in a PR for end2end testing ? Was there some issue with the progression that needed to be adressed to be testable in the end2end ?

**Review by Simmanz — APPROVED** [review]
### Review of the test

Great job on the end-to-end test! Your commentary in your end2end test, they ensure that a third party can easily know what is going on, which is incredibly helpful for others reviewing or updating this code in the future. The test also effectively cover thoroughly the user flow, ensuring that the app’s functionality is verified.

### Improvement for future PRs

I approve of this PR, but I did want to note some feedback for future reference. This PR was created specifically for end-to-end tests, yet it also includes commits for new UI features and other unrelated test files. While I understand these updates might be essential, it would be better to separate different features and testing changes into their respective PRs. Keeping each PR focused on a single scope (e.g., end-to-end tests only) helps maintain clarity, traceability, and faster reviews.

Thanks for your hard work !

File: app/src/androidTest/java/com/android/streetworkapp/ui/Progression/ProgressionTest.kt
Comment (Alex720p) [reply app/src/androidTest/java/com/android/streetworkapp/ui/Progression/ProgressionTest.kt:1]: I reffered to this one when I wrote: _Note: Ignore all files that are related to Progression, I had to merge the branch from pr https://github.com/SwEnt-Group8/Street-work-app/pull/93 into this one to integrate the end2end._
They'll dissapear once we merge my other branch

File: app/src/androidTest/java/com/android/streetworkapp/ui/navigation/BottomNavigationTest.kt
Comment (Alex720p) [reply app/src/androidTest/java/com/android/streetworkapp/ui/navigation/BottomNavigationTest.kt:1]: I probably shoudln't have. There was a bug in MainActivity that I fixed in this pr, this caused the Bottom and TopAppBar tests to fail. I fixed them in this pr but in retrospective I should have done them in another one

File: app/src/androidTest/java/com/android/streetworkapp/ui/navigation/TopAppBarTest.kt
Comment (Alex720p) [reply app/src/androidTest/java/com/android/streetworkapp/ui/navigation/TopAppBarTest.kt:1]: same here as for Bottom tests

File: app/src/main/java/com/android/streetworkapp/ui/progress/Progression.kt
Comment (Alex720p) [reply app/src/main/java/com/android/streetworkapp/ui/progress/Progression.kt:1]: I refered to this one when I wrote: Note: Ignore all files that are related to Progression, I had to merge the branch from pr https://github.com/SwEnt-Group8/Street-work-app/pull/93 into this one to integrate the end2end.
They'll dissapear once we merge my other branch

File: app/src/androidTest/java/com/android/streetworkapp/ui/Progression/ProgressionTest.kt [hunk 1]
```diff
@@ -1,35 +1,147 @@
...
+            achievements = emptyList())
+
+    val getProgressionOnSuccessSlot = slot<(Progression) -> Unit>()
+    every {
+      progressionRepository.getProgression(any(), capture(getProgressionOnSuccessSlot), any())
+    } answers { getProgressionOnSuccessSlot.captured(mockedProgression) }
+
+    composeTestRule.setContent {
+      ProgressScreen(navigationActions, userViewModel, progressionViewModel)
+    }
```
Comment (kzepfl) [inline app/src/androidTest/java/com/android/streetworkapp/ui/Progression/ProgressionTest.kt:143]:
```
composeTestRule.setContent {
      ProgressScreen(navigationActions, userViewModel, progressionViewModel)
    }
```
set content is used in all of the test you can put it in the @before setup() function
code in line : 58, 66, 95, 141

**Review by kzepfl — COMMENTED** [review]
this PR is quite big so I checked it aswell, I saw some code in test that could be changed. But apart from that I didn't see anything.

File: app/src/androidTest/java/com/android/streetworkapp/ui/Progression/ProgressionTest.kt (hunk 1 above)
Comment (Alex720p) [reply app/src/androidTest/java/com/android/streetworkapp/ui/Progression/ProgressionTest.kt:143]: This is not relevant to this pr:
_Note: Ignore all files that are related to Progression, I had to merge the branch from pr https://github.com/SwEnt-Group8/Street-work-app/pull/93 into this one to integrate the end2end._
I'll take into account for https://github.com/SwEnt-Group8/Street-work-app/pull/93

**Pull request merged**